<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head><meta http-equiv="content-type" content="text/html; charset=UTF-8" />
<title>CheckerPowerTest xref</title>
<link type="text/css" rel="stylesheet" href="../../../../../../stylesheet.css" />
</head>
<body>
<div id="overview"><a href="../../../../../../../testapidocs/com/puppycrawl/tools/checkstyle/internal/powermock/CheckerPowerTest.html">View Javadoc</a></div><pre>
<a class="jxr_linenumber" name="L1" href="#L1">1</a>   <em class="jxr_comment">////////////////////////////////////////////////////////////////////////////////</em>
<a class="jxr_linenumber" name="L2" href="#L2">2</a>   <em class="jxr_comment">// checkstyle: Checks Java source code for adherence to a set of rules.</em>
<a class="jxr_linenumber" name="L3" href="#L3">3</a>   <em class="jxr_comment">// Copyright (C) 2001-2019 the original author or authors.</em>
<a class="jxr_linenumber" name="L4" href="#L4">4</a>   <em class="jxr_comment">//</em>
<a class="jxr_linenumber" name="L5" href="#L5">5</a>   <em class="jxr_comment">// This library is free software; you can redistribute it and/or</em>
<a class="jxr_linenumber" name="L6" href="#L6">6</a>   <em class="jxr_comment">// modify it under the terms of the GNU Lesser General Public</em>
<a class="jxr_linenumber" name="L7" href="#L7">7</a>   <em class="jxr_comment">// License as published by the Free Software Foundation; either</em>
<a class="jxr_linenumber" name="L8" href="#L8">8</a>   <em class="jxr_comment">// version 2.1 of the License, or (at your option) any later version.</em>
<a class="jxr_linenumber" name="L9" href="#L9">9</a>   <em class="jxr_comment">//</em>
<a class="jxr_linenumber" name="L10" href="#L10">10</a>  <em class="jxr_comment">// This library is distributed in the hope that it will be useful,</em>
<a class="jxr_linenumber" name="L11" href="#L11">11</a>  <em class="jxr_comment">// but WITHOUT ANY WARRANTY; without even the implied warranty of</em>
<a class="jxr_linenumber" name="L12" href="#L12">12</a>  <em class="jxr_comment">// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU</em>
<a class="jxr_linenumber" name="L13" href="#L13">13</a>  <em class="jxr_comment">// Lesser General Public License for more details.</em>
<a class="jxr_linenumber" name="L14" href="#L14">14</a>  <em class="jxr_comment">//</em>
<a class="jxr_linenumber" name="L15" href="#L15">15</a>  <em class="jxr_comment">// You should have received a copy of the GNU Lesser General Public</em>
<a class="jxr_linenumber" name="L16" href="#L16">16</a>  <em class="jxr_comment">// License along with this library; if not, write to the Free Software</em>
<a class="jxr_linenumber" name="L17" href="#L17">17</a>  <em class="jxr_comment">// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA</em>
<a class="jxr_linenumber" name="L18" href="#L18">18</a>  <em class="jxr_comment">////////////////////////////////////////////////////////////////////////////////</em>
<a class="jxr_linenumber" name="L19" href="#L19">19</a>  
<a class="jxr_linenumber" name="L20" href="#L20">20</a>  <strong class="jxr_keyword">package</strong> com.puppycrawl.tools.checkstyle.internal.powermock;
<a class="jxr_linenumber" name="L21" href="#L21">21</a>  
<a class="jxr_linenumber" name="L22" href="#L22">22</a>  <strong class="jxr_keyword">import</strong> <strong class="jxr_keyword">static</strong> org.hamcrest.CoreMatchers.instanceOf;
<a class="jxr_linenumber" name="L23" href="#L23">23</a>  <strong class="jxr_keyword">import</strong> <strong class="jxr_keyword">static</strong> org.junit.Assert.assertEquals;
<a class="jxr_linenumber" name="L24" href="#L24">24</a>  <strong class="jxr_keyword">import</strong> <strong class="jxr_keyword">static</strong> org.junit.Assert.assertNull;
<a class="jxr_linenumber" name="L25" href="#L25">25</a>  <strong class="jxr_keyword">import</strong> <strong class="jxr_keyword">static</strong> org.junit.Assert.assertThat;
<a class="jxr_linenumber" name="L26" href="#L26">26</a>  <strong class="jxr_keyword">import</strong> <strong class="jxr_keyword">static</strong> org.junit.Assert.fail;
<a class="jxr_linenumber" name="L27" href="#L27">27</a>  <strong class="jxr_keyword">import</strong> <strong class="jxr_keyword">static</strong> org.mockito.Mockito.when;
<a class="jxr_linenumber" name="L28" href="#L28">28</a>  
<a class="jxr_linenumber" name="L29" href="#L29">29</a>  <strong class="jxr_keyword">import</strong> java.io.File;
<a class="jxr_linenumber" name="L30" href="#L30">30</a>  <strong class="jxr_keyword">import</strong> java.io.IOError;
<a class="jxr_linenumber" name="L31" href="#L31">31</a>  <strong class="jxr_keyword">import</strong> java.nio.charset.StandardCharsets;
<a class="jxr_linenumber" name="L32" href="#L32">32</a>  <strong class="jxr_keyword">import</strong> java.nio.file.Files;
<a class="jxr_linenumber" name="L33" href="#L33">33</a>  <strong class="jxr_keyword">import</strong> java.util.ArrayList;
<a class="jxr_linenumber" name="L34" href="#L34">34</a>  <strong class="jxr_keyword">import</strong> java.util.List;
<a class="jxr_linenumber" name="L35" href="#L35">35</a>  <strong class="jxr_keyword">import</strong> java.util.Properties;
<a class="jxr_linenumber" name="L36" href="#L36">36</a>  
<a class="jxr_linenumber" name="L37" href="#L37">37</a>  <strong class="jxr_keyword">import</strong> org.junit.Rule;
<a class="jxr_linenumber" name="L38" href="#L38">38</a>  <strong class="jxr_keyword">import</strong> org.junit.Test;
<a class="jxr_linenumber" name="L39" href="#L39">39</a>  <strong class="jxr_keyword">import</strong> org.junit.rules.TemporaryFolder;
<a class="jxr_linenumber" name="L40" href="#L40">40</a>  <strong class="jxr_keyword">import</strong> org.powermock.api.mockito.PowerMockito;
<a class="jxr_linenumber" name="L41" href="#L41">41</a>  
<a class="jxr_linenumber" name="L42" href="#L42">42</a>  <strong class="jxr_keyword">import</strong> com.puppycrawl.tools.checkstyle.Checker;
<a class="jxr_linenumber" name="L43" href="#L43">43</a>  <strong class="jxr_keyword">import</strong> com.puppycrawl.tools.checkstyle.DefaultConfiguration;
<a class="jxr_linenumber" name="L44" href="#L44">44</a>  <strong class="jxr_keyword">import</strong> com.puppycrawl.tools.checkstyle.api.CheckstyleException;
<a class="jxr_linenumber" name="L45" href="#L45">45</a>  
<a class="jxr_linenumber" name="L46" href="#L46">46</a>  <strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">class</strong> <a href="../../../../../../com/puppycrawl/tools/checkstyle/internal/powermock/CheckerPowerTest.html">CheckerPowerTest</a> {
<a class="jxr_linenumber" name="L47" href="#L47">47</a>  
<a class="jxr_linenumber" name="L48" href="#L48">48</a>      @Rule
<a class="jxr_linenumber" name="L49" href="#L49">49</a>      <strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">final</strong> TemporaryFolder temporaryFolder = <strong class="jxr_keyword">new</strong> TemporaryFolder();
<a class="jxr_linenumber" name="L50" href="#L50">50</a>  
<a class="jxr_linenumber" name="L51" href="#L51">51</a>      @Test
<a class="jxr_linenumber" name="L52" href="#L52">52</a>      <strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">void</strong> testCatchErrorInProcessFilesMethod() <strong class="jxr_keyword">throws</strong> Exception {
<a class="jxr_linenumber" name="L53" href="#L53">53</a>          <em class="jxr_comment">// The idea of the test is to satisfy coverage rate.</em>
<a class="jxr_linenumber" name="L54" href="#L54">54</a>          <em class="jxr_comment">// An Error indicates serious problems that a reasonable application should not try to</em>
<a class="jxr_linenumber" name="L55" href="#L55">55</a>          <em class="jxr_comment">// catch, but due to issue https://github.com/checkstyle/checkstyle/issues/2285</em>
<a class="jxr_linenumber" name="L56" href="#L56">56</a>          <em class="jxr_comment">// we catch errors in 'processFiles' method. Most such errors are abnormal conditions,</em>
<a class="jxr_linenumber" name="L57" href="#L57">57</a>          <em class="jxr_comment">// that is why we use PowerMockito to reproduce them.</em>
<a class="jxr_linenumber" name="L58" href="#L58">58</a>          <strong class="jxr_keyword">final</strong> File mock = PowerMockito.mock(File.<strong class="jxr_keyword">class</strong>);
<a class="jxr_linenumber" name="L59" href="#L59">59</a>          <em class="jxr_comment">// Assume that I/O error is happened when we try to invoke 'lastModified()' method.</em>
<a class="jxr_linenumber" name="L60" href="#L60">60</a>          <strong class="jxr_keyword">final</strong> String errorMessage = <span class="jxr_string">"Java Virtual Machine is broken"</span>
<a class="jxr_linenumber" name="L61" href="#L61">61</a>              + <span class="jxr_string">" or has run out of resources necessary for it to continue operating."</span>;
<a class="jxr_linenumber" name="L62" href="#L62">62</a>          <strong class="jxr_keyword">final</strong> Error expectedError = <strong class="jxr_keyword">new</strong> IOError(<strong class="jxr_keyword">new</strong> InternalError(errorMessage));
<a class="jxr_linenumber" name="L63" href="#L63">63</a>          when(mock.lastModified()).thenThrow(expectedError);
<a class="jxr_linenumber" name="L64" href="#L64">64</a>          when(mock.getAbsolutePath()).thenReturn(<span class="jxr_string">"testFile"</span>);
<a class="jxr_linenumber" name="L65" href="#L65">65</a>          <strong class="jxr_keyword">final</strong> Checker checker = <strong class="jxr_keyword">new</strong> Checker();
<a class="jxr_linenumber" name="L66" href="#L66">66</a>          <strong class="jxr_keyword">final</strong> List&lt;File&gt; filesToProcess = <strong class="jxr_keyword">new</strong> ArrayList&lt;&gt;();
<a class="jxr_linenumber" name="L67" href="#L67">67</a>          filesToProcess.add(mock);
<a class="jxr_linenumber" name="L68" href="#L68">68</a>          <strong class="jxr_keyword">try</strong> {
<a class="jxr_linenumber" name="L69" href="#L69">69</a>              checker.process(filesToProcess);
<a class="jxr_linenumber" name="L70" href="#L70">70</a>              fail(<span class="jxr_string">"IOError is expected!"</span>);
<a class="jxr_linenumber" name="L71" href="#L71">71</a>          }
<a class="jxr_linenumber" name="L72" href="#L72">72</a>          <em class="jxr_comment">// -@cs[IllegalCatchExtended] Testing for catch Error is part of 100% coverage.</em>
<a class="jxr_linenumber" name="L73" href="#L73">73</a>          <strong class="jxr_keyword">catch</strong> (Error error) {
<a class="jxr_linenumber" name="L74" href="#L74">74</a>              assertThat(<span class="jxr_string">"Error cause differs from IOError"</span>,
<a class="jxr_linenumber" name="L75" href="#L75">75</a>                      error.getCause(), instanceOf(IOError.<strong class="jxr_keyword">class</strong>));
<a class="jxr_linenumber" name="L76" href="#L76">76</a>              assertThat(<span class="jxr_string">"Error cause is not InternalError"</span>,
<a class="jxr_linenumber" name="L77" href="#L77">77</a>                      error.getCause().getCause(), instanceOf(InternalError.<strong class="jxr_keyword">class</strong>));
<a class="jxr_linenumber" name="L78" href="#L78">78</a>              assertEquals(<span class="jxr_string">"Error message is not expected"</span>,
<a class="jxr_linenumber" name="L79" href="#L79">79</a>                      errorMessage, error.getCause().getCause().getMessage());
<a class="jxr_linenumber" name="L80" href="#L80">80</a>          }
<a class="jxr_linenumber" name="L81" href="#L81">81</a>      }
<a class="jxr_linenumber" name="L82" href="#L82">82</a>  
<a class="jxr_linenumber" name="L83" href="#L83">83</a>      @Test
<a class="jxr_linenumber" name="L84" href="#L84">84</a>      <strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">void</strong> testCatchErrorWithNoFileName() <strong class="jxr_keyword">throws</strong> Exception {
<a class="jxr_linenumber" name="L85" href="#L85">85</a>          <em class="jxr_comment">// The idea of the test is to satisfy coverage rate.</em>
<a class="jxr_linenumber" name="L86" href="#L86">86</a>          <em class="jxr_comment">// An Error indicates serious problems that a reasonable application should not try to</em>
<a class="jxr_linenumber" name="L87" href="#L87">87</a>          <em class="jxr_comment">// catch, but due to issue https://github.com/checkstyle/checkstyle/issues/2285</em>
<a class="jxr_linenumber" name="L88" href="#L88">88</a>          <em class="jxr_comment">// we catch errors in 'processFiles' method. Most such errors are abnormal conditions,</em>
<a class="jxr_linenumber" name="L89" href="#L89">89</a>          <em class="jxr_comment">// that is why we use PowerMockito to reproduce them.</em>
<a class="jxr_linenumber" name="L90" href="#L90">90</a>          <strong class="jxr_keyword">final</strong> File mock = PowerMockito.mock(File.<strong class="jxr_keyword">class</strong>);
<a class="jxr_linenumber" name="L91" href="#L91">91</a>          <em class="jxr_comment">// Assume that I/O error is happened when we try to invoke 'lastModified()' method.</em>
<a class="jxr_linenumber" name="L92" href="#L92">92</a>          <strong class="jxr_keyword">final</strong> String errorMessage = <span class="jxr_string">"Java Virtual Machine is broken"</span>
<a class="jxr_linenumber" name="L93" href="#L93">93</a>              + <span class="jxr_string">" or has run out of resources necessary for it to continue operating."</span>;
<a class="jxr_linenumber" name="L94" href="#L94">94</a>          <strong class="jxr_keyword">final</strong> Error expectedError = <strong class="jxr_keyword">new</strong> IOError(<strong class="jxr_keyword">new</strong> InternalError(errorMessage));
<a class="jxr_linenumber" name="L95" href="#L95">95</a>          when(mock.lastModified()).thenThrow(expectedError);
<a class="jxr_linenumber" name="L96" href="#L96">96</a>          <strong class="jxr_keyword">final</strong> Checker checker = <strong class="jxr_keyword">new</strong> Checker();
<a class="jxr_linenumber" name="L97" href="#L97">97</a>          <strong class="jxr_keyword">final</strong> List&lt;File&gt; filesToProcess = <strong class="jxr_keyword">new</strong> ArrayList&lt;&gt;();
<a class="jxr_linenumber" name="L98" href="#L98">98</a>          filesToProcess.add(mock);
<a class="jxr_linenumber" name="L99" href="#L99">99</a>          <strong class="jxr_keyword">try</strong> {
<a class="jxr_linenumber" name="L100" href="#L100">100</a>             checker.process(filesToProcess);
<a class="jxr_linenumber" name="L101" href="#L101">101</a>             fail(<span class="jxr_string">"IOError is expected!"</span>);
<a class="jxr_linenumber" name="L102" href="#L102">102</a>         }
<a class="jxr_linenumber" name="L103" href="#L103">103</a>         <em class="jxr_comment">// -@cs[IllegalCatchExtended] Testing for catch Error is part of 100% coverage.</em>
<a class="jxr_linenumber" name="L104" href="#L104">104</a>         <strong class="jxr_keyword">catch</strong> (Error error) {
<a class="jxr_linenumber" name="L105" href="#L105">105</a>             assertThat(<span class="jxr_string">"Error cause differs from IOError"</span>,
<a class="jxr_linenumber" name="L106" href="#L106">106</a>                     error.getCause(), instanceOf(IOError.<strong class="jxr_keyword">class</strong>));
<a class="jxr_linenumber" name="L107" href="#L107">107</a>             assertThat(<span class="jxr_string">"Error cause is not InternalError"</span>,
<a class="jxr_linenumber" name="L108" href="#L108">108</a>                     error.getCause().getCause(), instanceOf(InternalError.<strong class="jxr_keyword">class</strong>));
<a class="jxr_linenumber" name="L109" href="#L109">109</a>             assertEquals(<span class="jxr_string">"Error message is not expected"</span>,
<a class="jxr_linenumber" name="L110" href="#L110">110</a>                     errorMessage, error.getCause().getCause().getMessage());
<a class="jxr_linenumber" name="L111" href="#L111">111</a>         }
<a class="jxr_linenumber" name="L112" href="#L112">112</a>     }
<a class="jxr_linenumber" name="L113" href="#L113">113</a> 
<a class="jxr_linenumber" name="L114" href="#L114">114</a>     @Test
<a class="jxr_linenumber" name="L115" href="#L115">115</a>     <strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">void</strong> testCatchErrorWithCache() <strong class="jxr_keyword">throws</strong> Exception {
<a class="jxr_linenumber" name="L116" href="#L116">116</a>         <strong class="jxr_keyword">final</strong> File cacheFile = temporaryFolder.newFile();
<a class="jxr_linenumber" name="L117" href="#L117">117</a> 
<a class="jxr_linenumber" name="L118" href="#L118">118</a>         <strong class="jxr_keyword">final</strong> DefaultConfiguration checkerConfig = <strong class="jxr_keyword">new</strong> DefaultConfiguration(<span class="jxr_string">"configuration"</span>);
<a class="jxr_linenumber" name="L119" href="#L119">119</a>         checkerConfig.addAttribute(<span class="jxr_string">"charset"</span>, StandardCharsets.UTF_8.name());
<a class="jxr_linenumber" name="L120" href="#L120">120</a>         checkerConfig.addAttribute(<span class="jxr_string">"cacheFile"</span>, cacheFile.getPath());
<a class="jxr_linenumber" name="L121" href="#L121">121</a> 
<a class="jxr_linenumber" name="L122" href="#L122">122</a>         <strong class="jxr_keyword">final</strong> File mock = PowerMockito.mock(File.<strong class="jxr_keyword">class</strong>);
<a class="jxr_linenumber" name="L123" href="#L123">123</a>         <strong class="jxr_keyword">final</strong> String errorMessage = <span class="jxr_string">"Java Virtual Machine is broken"</span>
<a class="jxr_linenumber" name="L124" href="#L124">124</a>             + <span class="jxr_string">" or has run out of resources necessary for it to continue operating."</span>;
<a class="jxr_linenumber" name="L125" href="#L125">125</a>         <strong class="jxr_keyword">final</strong> Error expectedError = <strong class="jxr_keyword">new</strong> IOError(<strong class="jxr_keyword">new</strong> InternalError(errorMessage));
<a class="jxr_linenumber" name="L126" href="#L126">126</a>         when(mock.getAbsolutePath()).thenReturn(<span class="jxr_string">"testFile"</span>);
<a class="jxr_linenumber" name="L127" href="#L127">127</a>         when(mock.getAbsoluteFile()).thenThrow(expectedError);
<a class="jxr_linenumber" name="L128" href="#L128">128</a>         <strong class="jxr_keyword">final</strong> Checker checker = <strong class="jxr_keyword">new</strong> Checker();
<a class="jxr_linenumber" name="L129" href="#L129">129</a>         checker.setModuleClassLoader(Thread.currentThread().getContextClassLoader());
<a class="jxr_linenumber" name="L130" href="#L130">130</a>         checker.configure(checkerConfig);
<a class="jxr_linenumber" name="L131" href="#L131">131</a>         <strong class="jxr_keyword">final</strong> List&lt;File&gt; filesToProcess = <strong class="jxr_keyword">new</strong> ArrayList&lt;&gt;();
<a class="jxr_linenumber" name="L132" href="#L132">132</a>         filesToProcess.add(mock);
<a class="jxr_linenumber" name="L133" href="#L133">133</a>         <strong class="jxr_keyword">try</strong> {
<a class="jxr_linenumber" name="L134" href="#L134">134</a>             checker.process(filesToProcess);
<a class="jxr_linenumber" name="L135" href="#L135">135</a>             fail(<span class="jxr_string">"IOError is expected!"</span>);
<a class="jxr_linenumber" name="L136" href="#L136">136</a>         }
<a class="jxr_linenumber" name="L137" href="#L137">137</a>         <em class="jxr_comment">// -@cs[IllegalCatchExtended] Testing for catch Error is part of 100% coverage.</em>
<a class="jxr_linenumber" name="L138" href="#L138">138</a>         <strong class="jxr_keyword">catch</strong> (Error error) {
<a class="jxr_linenumber" name="L139" href="#L139">139</a>             assertThat(<span class="jxr_string">"Error cause differs from IOError"</span>,
<a class="jxr_linenumber" name="L140" href="#L140">140</a>                     error.getCause(), instanceOf(IOError.<strong class="jxr_keyword">class</strong>));
<a class="jxr_linenumber" name="L141" href="#L141">141</a>             assertEquals(<span class="jxr_string">"Error message is not expected"</span>,
<a class="jxr_linenumber" name="L142" href="#L142">142</a>                     errorMessage, error.getCause().getCause().getMessage());
<a class="jxr_linenumber" name="L143" href="#L143">143</a> 
<a class="jxr_linenumber" name="L144" href="#L144">144</a>             <em class="jxr_comment">// destroy is called by Main</em>
<a class="jxr_linenumber" name="L145" href="#L145">145</a>             checker.destroy();
<a class="jxr_linenumber" name="L146" href="#L146">146</a> 
<a class="jxr_linenumber" name="L147" href="#L147">147</a>             <strong class="jxr_keyword">final</strong> Properties cache = <strong class="jxr_keyword">new</strong> Properties();
<a class="jxr_linenumber" name="L148" href="#L148">148</a>             cache.load(Files.newBufferedReader(cacheFile.toPath()));
<a class="jxr_linenumber" name="L149" href="#L149">149</a> 
<a class="jxr_linenumber" name="L150" href="#L150">150</a>             assertEquals(<span class="jxr_string">"Cache has unexpected size"</span>,
<a class="jxr_linenumber" name="L151" href="#L151">151</a>                     1, cache.size());
<a class="jxr_linenumber" name="L152" href="#L152">152</a>             assertNull(<span class="jxr_string">"testFile is not in cache"</span>,
<a class="jxr_linenumber" name="L153" href="#L153">153</a>                     cache.getProperty(<span class="jxr_string">"testFile"</span>));
<a class="jxr_linenumber" name="L154" href="#L154">154</a>         }
<a class="jxr_linenumber" name="L155" href="#L155">155</a>     }
<a class="jxr_linenumber" name="L156" href="#L156">156</a> 
<a class="jxr_linenumber" name="L157" href="#L157">157</a>     @Test
<a class="jxr_linenumber" name="L158" href="#L158">158</a>     <strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">void</strong> testCatchErrorWithCacheWithNoFileName() <strong class="jxr_keyword">throws</strong> Exception {
<a class="jxr_linenumber" name="L159" href="#L159">159</a>         <strong class="jxr_keyword">final</strong> File cacheFile = temporaryFolder.newFile();
<a class="jxr_linenumber" name="L160" href="#L160">160</a> 
<a class="jxr_linenumber" name="L161" href="#L161">161</a>         <strong class="jxr_keyword">final</strong> DefaultConfiguration checkerConfig = <strong class="jxr_keyword">new</strong> DefaultConfiguration(<span class="jxr_string">"configuration"</span>);
<a class="jxr_linenumber" name="L162" href="#L162">162</a>         checkerConfig.addAttribute(<span class="jxr_string">"charset"</span>, StandardCharsets.UTF_8.name());
<a class="jxr_linenumber" name="L163" href="#L163">163</a>         checkerConfig.addAttribute(<span class="jxr_string">"cacheFile"</span>, cacheFile.getPath());
<a class="jxr_linenumber" name="L164" href="#L164">164</a> 
<a class="jxr_linenumber" name="L165" href="#L165">165</a>         <strong class="jxr_keyword">final</strong> File mock = PowerMockito.mock(File.<strong class="jxr_keyword">class</strong>);
<a class="jxr_linenumber" name="L166" href="#L166">166</a>         <strong class="jxr_keyword">final</strong> String errorMessage = <span class="jxr_string">"Java Virtual Machine is broken"</span>
<a class="jxr_linenumber" name="L167" href="#L167">167</a>             + <span class="jxr_string">" or has run out of resources necessary for it to continue operating."</span>;
<a class="jxr_linenumber" name="L168" href="#L168">168</a>         <strong class="jxr_keyword">final</strong> Error expectedError = <strong class="jxr_keyword">new</strong> IOError(<strong class="jxr_keyword">new</strong> InternalError(errorMessage));
<a class="jxr_linenumber" name="L169" href="#L169">169</a>         when(mock.getAbsolutePath()).thenThrow(expectedError);
<a class="jxr_linenumber" name="L170" href="#L170">170</a>         <strong class="jxr_keyword">final</strong> Checker checker = <strong class="jxr_keyword">new</strong> Checker();
<a class="jxr_linenumber" name="L171" href="#L171">171</a>         checker.setModuleClassLoader(Thread.currentThread().getContextClassLoader());
<a class="jxr_linenumber" name="L172" href="#L172">172</a>         checker.configure(checkerConfig);
<a class="jxr_linenumber" name="L173" href="#L173">173</a>         <strong class="jxr_keyword">final</strong> List&lt;File&gt; filesToProcess = <strong class="jxr_keyword">new</strong> ArrayList&lt;&gt;();
<a class="jxr_linenumber" name="L174" href="#L174">174</a>         filesToProcess.add(mock);
<a class="jxr_linenumber" name="L175" href="#L175">175</a>         <strong class="jxr_keyword">try</strong> {
<a class="jxr_linenumber" name="L176" href="#L176">176</a>             checker.process(filesToProcess);
<a class="jxr_linenumber" name="L177" href="#L177">177</a>             fail(<span class="jxr_string">"IOError is expected!"</span>);
<a class="jxr_linenumber" name="L178" href="#L178">178</a>         }
<a class="jxr_linenumber" name="L179" href="#L179">179</a>         <em class="jxr_comment">// -@cs[IllegalCatchExtended] Testing for catch Error is part of 100% coverage.</em>
<a class="jxr_linenumber" name="L180" href="#L180">180</a>         <strong class="jxr_keyword">catch</strong> (Error error) {
<a class="jxr_linenumber" name="L181" href="#L181">181</a>             assertThat(<span class="jxr_string">"Error cause differs from IOError"</span>,
<a class="jxr_linenumber" name="L182" href="#L182">182</a>                     error.getCause(), instanceOf(IOError.<strong class="jxr_keyword">class</strong>));
<a class="jxr_linenumber" name="L183" href="#L183">183</a>             assertEquals(<span class="jxr_string">"Error message is not expected"</span>,
<a class="jxr_linenumber" name="L184" href="#L184">184</a>                     errorMessage, error.getCause().getCause().getMessage());
<a class="jxr_linenumber" name="L185" href="#L185">185</a> 
<a class="jxr_linenumber" name="L186" href="#L186">186</a>             <em class="jxr_comment">// destroy is called by Main</em>
<a class="jxr_linenumber" name="L187" href="#L187">187</a>             checker.destroy();
<a class="jxr_linenumber" name="L188" href="#L188">188</a> 
<a class="jxr_linenumber" name="L189" href="#L189">189</a>             <strong class="jxr_keyword">final</strong> Properties cache = <strong class="jxr_keyword">new</strong> Properties();
<a class="jxr_linenumber" name="L190" href="#L190">190</a>             cache.load(Files.newBufferedReader(cacheFile.toPath()));
<a class="jxr_linenumber" name="L191" href="#L191">191</a> 
<a class="jxr_linenumber" name="L192" href="#L192">192</a>             assertEquals(<span class="jxr_string">"Cache has unexpected size"</span>,
<a class="jxr_linenumber" name="L193" href="#L193">193</a>                     1, cache.size());
<a class="jxr_linenumber" name="L194" href="#L194">194</a>         }
<a class="jxr_linenumber" name="L195" href="#L195">195</a>     }
<a class="jxr_linenumber" name="L196" href="#L196">196</a> 
<a class="jxr_linenumber" name="L197" href="#L197">197</a>     @Test
<a class="jxr_linenumber" name="L198" href="#L198">198</a>     <strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">void</strong> testExceptionWithNoFileName() {
<a class="jxr_linenumber" name="L199" href="#L199">199</a>         <em class="jxr_comment">// The idea of the test is to satisfy coverage rate.</em>
<a class="jxr_linenumber" name="L200" href="#L200">200</a>         <em class="jxr_comment">// An Error indicates serious problems that a reasonable application should not try to</em>
<a class="jxr_linenumber" name="L201" href="#L201">201</a>         <em class="jxr_comment">// catch, but due to issue https://github.com/checkstyle/checkstyle/issues/2285</em>
<a class="jxr_linenumber" name="L202" href="#L202">202</a>         <em class="jxr_comment">// we catch errors in 'processFiles' method. Most such errors are abnormal conditions,</em>
<a class="jxr_linenumber" name="L203" href="#L203">203</a>         <em class="jxr_comment">// that is why we use PowerMockito to reproduce them.</em>
<a class="jxr_linenumber" name="L204" href="#L204">204</a>         <strong class="jxr_keyword">final</strong> File mock = PowerMockito.mock(File.<strong class="jxr_keyword">class</strong>);
<a class="jxr_linenumber" name="L205" href="#L205">205</a>         <strong class="jxr_keyword">final</strong> String errorMessage = <span class="jxr_string">"Security Exception"</span>;
<a class="jxr_linenumber" name="L206" href="#L206">206</a>         <strong class="jxr_keyword">final</strong> Exception expectedError = <strong class="jxr_keyword">new</strong> SecurityException(errorMessage);
<a class="jxr_linenumber" name="L207" href="#L207">207</a>         when(mock.getAbsolutePath()).thenThrow(expectedError);
<a class="jxr_linenumber" name="L208" href="#L208">208</a>         <strong class="jxr_keyword">final</strong> Checker checker = <strong class="jxr_keyword">new</strong> Checker();
<a class="jxr_linenumber" name="L209" href="#L209">209</a>         <strong class="jxr_keyword">final</strong> List&lt;File&gt; filesToProcess = <strong class="jxr_keyword">new</strong> ArrayList&lt;&gt;();
<a class="jxr_linenumber" name="L210" href="#L210">210</a>         filesToProcess.add(mock);
<a class="jxr_linenumber" name="L211" href="#L211">211</a>         <strong class="jxr_keyword">try</strong> {
<a class="jxr_linenumber" name="L212" href="#L212">212</a>             checker.process(filesToProcess);
<a class="jxr_linenumber" name="L213" href="#L213">213</a>             fail(<span class="jxr_string">"IOError is expected!"</span>);
<a class="jxr_linenumber" name="L214" href="#L214">214</a>         }
<a class="jxr_linenumber" name="L215" href="#L215">215</a>         <strong class="jxr_keyword">catch</strong> (CheckstyleException ex) {
<a class="jxr_linenumber" name="L216" href="#L216">216</a>             assertThat(<span class="jxr_string">"Error cause differs from SecurityException"</span>,
<a class="jxr_linenumber" name="L217" href="#L217">217</a>                     ex.getCause(), instanceOf(SecurityException.<strong class="jxr_keyword">class</strong>));
<a class="jxr_linenumber" name="L218" href="#L218">218</a>             assertEquals(<span class="jxr_string">"Error message is not expected"</span>,
<a class="jxr_linenumber" name="L219" href="#L219">219</a>                     errorMessage, ex.getCause().getMessage());
<a class="jxr_linenumber" name="L220" href="#L220">220</a>         }
<a class="jxr_linenumber" name="L221" href="#L221">221</a>     }
<a class="jxr_linenumber" name="L222" href="#L222">222</a> 
<a class="jxr_linenumber" name="L223" href="#L223">223</a>     @Test
<a class="jxr_linenumber" name="L224" href="#L224">224</a>     <strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">void</strong> testExceptionWithCacheAndNoFileName() <strong class="jxr_keyword">throws</strong> Exception {
<a class="jxr_linenumber" name="L225" href="#L225">225</a>         <strong class="jxr_keyword">final</strong> File cacheFile = temporaryFolder.newFile();
<a class="jxr_linenumber" name="L226" href="#L226">226</a> 
<a class="jxr_linenumber" name="L227" href="#L227">227</a>         <strong class="jxr_keyword">final</strong> DefaultConfiguration checkerConfig = <strong class="jxr_keyword">new</strong> DefaultConfiguration(<span class="jxr_string">"configuration"</span>);
<a class="jxr_linenumber" name="L228" href="#L228">228</a>         checkerConfig.addAttribute(<span class="jxr_string">"charset"</span>, StandardCharsets.UTF_8.name());
<a class="jxr_linenumber" name="L229" href="#L229">229</a>         checkerConfig.addAttribute(<span class="jxr_string">"cacheFile"</span>, cacheFile.getPath());
<a class="jxr_linenumber" name="L230" href="#L230">230</a> 
<a class="jxr_linenumber" name="L231" href="#L231">231</a>         <em class="jxr_comment">// The idea of the test is to satisfy coverage rate.</em>
<a class="jxr_linenumber" name="L232" href="#L232">232</a>         <em class="jxr_comment">// An Error indicates serious problems that a reasonable application should not try to</em>
<a class="jxr_linenumber" name="L233" href="#L233">233</a>         <em class="jxr_comment">// catch, but due to issue https://github.com/checkstyle/checkstyle/issues/2285</em>
<a class="jxr_linenumber" name="L234" href="#L234">234</a>         <em class="jxr_comment">// we catch errors in 'processFiles' method. Most such errors are abnormal conditions,</em>
<a class="jxr_linenumber" name="L235" href="#L235">235</a>         <em class="jxr_comment">// that is why we use PowerMockito to reproduce them.</em>
<a class="jxr_linenumber" name="L236" href="#L236">236</a>         <strong class="jxr_keyword">final</strong> File mock = PowerMockito.mock(File.<strong class="jxr_keyword">class</strong>);
<a class="jxr_linenumber" name="L237" href="#L237">237</a>         <strong class="jxr_keyword">final</strong> String errorMessage = <span class="jxr_string">"Security Exception"</span>;
<a class="jxr_linenumber" name="L238" href="#L238">238</a>         <strong class="jxr_keyword">final</strong> Exception expectedError = <strong class="jxr_keyword">new</strong> SecurityException(errorMessage);
<a class="jxr_linenumber" name="L239" href="#L239">239</a>         when(mock.getAbsolutePath()).thenThrow(expectedError);
<a class="jxr_linenumber" name="L240" href="#L240">240</a>         <strong class="jxr_keyword">final</strong> Checker checker = <strong class="jxr_keyword">new</strong> Checker();
<a class="jxr_linenumber" name="L241" href="#L241">241</a>         checker.setModuleClassLoader(Thread.currentThread().getContextClassLoader());
<a class="jxr_linenumber" name="L242" href="#L242">242</a>         checker.configure(checkerConfig);
<a class="jxr_linenumber" name="L243" href="#L243">243</a>         <strong class="jxr_keyword">final</strong> List&lt;File&gt; filesToProcess = <strong class="jxr_keyword">new</strong> ArrayList&lt;&gt;();
<a class="jxr_linenumber" name="L244" href="#L244">244</a>         filesToProcess.add(mock);
<a class="jxr_linenumber" name="L245" href="#L245">245</a>         <strong class="jxr_keyword">try</strong> {
<a class="jxr_linenumber" name="L246" href="#L246">246</a>             checker.process(filesToProcess);
<a class="jxr_linenumber" name="L247" href="#L247">247</a>             fail(<span class="jxr_string">"IOError is expected!"</span>);
<a class="jxr_linenumber" name="L248" href="#L248">248</a>         }
<a class="jxr_linenumber" name="L249" href="#L249">249</a>         <strong class="jxr_keyword">catch</strong> (CheckstyleException ex) {
<a class="jxr_linenumber" name="L250" href="#L250">250</a>             assertThat(<span class="jxr_string">"Error cause differs from SecurityException"</span>,
<a class="jxr_linenumber" name="L251" href="#L251">251</a>                     ex.getCause(), instanceOf(SecurityException.<strong class="jxr_keyword">class</strong>));
<a class="jxr_linenumber" name="L252" href="#L252">252</a>             assertEquals(<span class="jxr_string">"Error message is not expected"</span>,
<a class="jxr_linenumber" name="L253" href="#L253">253</a>                     errorMessage, ex.getCause().getMessage());
<a class="jxr_linenumber" name="L254" href="#L254">254</a> 
<a class="jxr_linenumber" name="L255" href="#L255">255</a>             <em class="jxr_comment">// destroy is called by Main</em>
<a class="jxr_linenumber" name="L256" href="#L256">256</a>             checker.destroy();
<a class="jxr_linenumber" name="L257" href="#L257">257</a> 
<a class="jxr_linenumber" name="L258" href="#L258">258</a>             <strong class="jxr_keyword">final</strong> Properties cache = <strong class="jxr_keyword">new</strong> Properties();
<a class="jxr_linenumber" name="L259" href="#L259">259</a>             cache.load(Files.newBufferedReader(cacheFile.toPath()));
<a class="jxr_linenumber" name="L260" href="#L260">260</a> 
<a class="jxr_linenumber" name="L261" href="#L261">261</a>             assertEquals(<span class="jxr_string">"Cache has unexpected size"</span>,
<a class="jxr_linenumber" name="L262" href="#L262">262</a>                     1, cache.size());
<a class="jxr_linenumber" name="L263" href="#L263">263</a>         }
<a class="jxr_linenumber" name="L264" href="#L264">264</a>     }
<a class="jxr_linenumber" name="L265" href="#L265">265</a> 
<a class="jxr_linenumber" name="L266" href="#L266">266</a> }
</pre>
<hr/>
<div id="footer">Copyright &#169; 2001&#x2013;2019. All rights reserved.</div>
</body>
</html>
